### PR TITLE
feat: add max transfer feature with feeEstimation

### PR DESCRIPTION
**Pull request by dhruvkelawala**

### CODE REVIEW

**Comment by dhruvkelawala** [comment]
Still in draft. Need to figure out various cases to handle:

1. Subtract maxFee from MAX amount only in case of ETH
2. How to handle negative values. i.e. Clicking on MAX when balance is less than maxFee. Currently, it sets a negative value obtained from `balance - maxFee` and disables `Send` button. Should this be handled more graciously?
3. Should we hide MAX button once it's clicked? i.e. `inputAmount === (balance - maxFee)` and show again when maxFee is updated and it's different from the previous maxFee? We can use [usePrevious()](https://usehooks.com/usePrevious/#:~:text=useHooks(%F0%9F%90%A0)&text=Hooks%20are%20a%20feature%20in,them%20in%20your%20next%20project.) hook for that.

**Comment by justingruss** [comment]
2. Don't think it's a good UX to show a negative value. Instead if user clicks 'MAX' when insufficient funds to cover maxFee, should populate the field with the max value user has available but also display error state with messaging explaining why they cannot proceed. Send button should be disabled

3. In updated UI the 'MAX' button will be hidden and fiat value will be shown in its place. Didn't account for changing maxFee. Maybe we could update input value automatically? On current UI we can keep 'MAX' button visible after clicking

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 1]
```diff
@@ -111,6 +151,64 @@ const SendSchema: Schema<SendInput> = object().required().shape({
...
+    throw new Error("Account, TokenAddress and Balance are required")
+  }
+
+  const { compileCalldata, estimatedFeeToMaxFee: addOverheadToFee } = stark
+
+  const call: Call = {
+    contractAddress: tokenAddress ?? "",
+    entrypoint: "transfer",
+    calldata: compileCalldata({
+      recipient: account?.address ?? "0",
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:173]: same

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 2]
```diff
@@ -111,6 +151,64 @@ const SendSchema: Schema<SendInput> = object().required().shape({
...
+  } = useSWR(
+    [
+      "maxEthTransferEstimate",
+      Math.floor(Date.now() / 60e3),
+      account.networkId,
+    ],
+    () => getEstimatedFee(call),
+    {
+      suspense: false,
+      refreshInterval: 15 * 1000,
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:191]:
```suggestion
      refreshInterval: 15e3,
```

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 3]
```diff
@@ -111,16 +151,79 @@ const SendSchema: Schema<SendInput> = object().required().shape({
...
+export const useMaxFeeEstimateForTransfer = (
+  tokenAddress?: string,
+  balance?: BigNumber,
+  account?: Account,
+): {
+  maxFee?: BigNumber
+  error?: any
+  loading: boolean
+} => {
+  if (!account || !balance || !tokenAddress) {
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:163]: check if token is the fee token, otherwise you can return a 0

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 4]
```diff
@@ -129,17 +232,65 @@ export const TokenScreen: FC = () => {
...
+  const handleMaxClick = async () => {
+    setMaxClicked(true)
+    setMaxInputAmount(token, maxFee)
+  }
+
+  const disableSubmit =
+    isSubmitting ||
+    (submitCount > 0 && !isDirty) ||
+    !inputAmount ||
+    !inputRecipient ||
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:290]: I would remove those as that gets handled by validation

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 5]
```diff
@@ -129,17 +232,65 @@ export const TokenScreen: FC = () => {
...
+    inputAmount > balance ||
+    parseFloat(inputAmount) < 0
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:292]: should be part of validation (yup schema)

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 6]
```diff
@@ -175,21 +326,37 @@ export const TokenScreen: FC = () => {
             navigate(routes.accountTokens())
           })}
         >
-          <ControlledInputText
+          <StyledControlledInput
             autoComplete="off"
-            control={control}
+            control={control as any}
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:331]: why is that needed?

Comment (dhruvkelawala) [reply packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:331]: I was adding better typings to `ControlledInputTextAlt` but didn't work. Maybe we should add typings to it.

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 7]
```diff
@@ -129,17 +248,60 @@ export const TokenScreen: FC = () => {
...
 
+  const {
+    maxFee,
+    error: maxFeeError,
+    loading: maxFeeLoading,
+  } = useMaxFeeEstimateForTransfer(token?.address, token?.balance, account)
+
+  useEffect(() => {
+    if (maxClicked && maxFee && token) {
+      setMaxInputAmount(token, maxFee)
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:267]: should be reset at some point
maybe reset if the user edits the value the next time

or just disable input after max click, with some option to reenable it while reseting this value

Comment (dhruvkelawala) [reply packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:268]: Hmm, makes sense

Comment (dhruvkelawala) [reply packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:268]: Done

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx [hunk 8]
```diff
@@ -129,17 +249,67 @@ export const TokenScreen: FC = () => {
...
+    if (token.balance && maxFee) {
+      const balanceBn = token.balance
+
+      const maxAmount = balanceBn.sub(maxFee.toString())
+
+      const formattedMaxAmount = utils.formatUnits(
+        maxAmount.toString(),
+        tokenDecimals,
+      )
+      setValue("amount", maxAmount.lte(0) ? tokenBalance : formattedMaxAmount)
```
Comment (janek26) [inline packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:288]: you can use BN.max with your value and `0`

and why would we input the tokenbalance in case the user can not send anyways? we should probably block him here already

**Review by janek26 — COMMENTED** [review]
one more thought:
what happens if I as a User that want to send as much as possible click the "Max button" and realize you dont put all of my balance in the field. If I now manually put my token balance in the value field, would it work? I think we should block people from forwarding to the next screen in case `token balance + fee > input amount`

**Comment by dhruvkelawala** [comment]
> f I now manually put my token balance in the value field

Not allowed case should be `updatedInputAmount + fee > balance`

File: packages/extension/src/ui/features/accountTokens/TokenScreen.tsx (hunk 8 above)
Comment (dhruvkelawala) [reply packages/extension/src/ui/features/accountTokens/TokenScreen.tsx:288]: We are blocking it with an error message as suggested by @justingruss in point 1

**Comment by dhruvkelawala** [comment]
Ready to merge

File: packages/extension/src/ui/components/InputText.tsx [hunk 9]
```diff
@@ -118,6 +187,61 @@ export const ControlledInputText = styled(
...
+          style={{ position: "relative" }}
+          {...props}
+          value={value || ""}
+          {...field}
+          inputRef={ref}
+          inputMode="decimal"
+          type="text"
+          onChange={(e: React.ChangeEvent<HTMLInputElement>) => {
+            console.log(
+              "🚀 ~ file: InputText.tsx ~ line 218 ~ e",
```
Comment (janek26) [inline packages/extension/src/ui/components/InputText.tsx:210]: console.log


and let me know which viscose plugin this is

**Review by janek26 — APPROVED** [review]
looks good!

File: packages/extension/src/ui/components/InputText.tsx (hunk 9 above)
Comment (dhruvkelawala) [reply packages/extension/src/ui/components/InputText.tsx]: [Tubro Console Log](https://marketplace.visualstudio.com/items?itemName=ChakrounAnas.turbo-console-log)
